fx quant: fix using size of quant layer in torch._assert

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53875 fx quant: remove unnecessary dequants for getattr
* #53614 fx quant: clean up names of quantize handlers
* #53196 fx quant: ensure fp32 sum works correctly with quantized layers
* **#53187 fx quant: fix using size of quant layer in torch._assert**
* #53120 fx quant: support for ndim followed by binary op

Summary:

Before this diff, if we had code lik

```
x = any_quant_layer(...)
x_size0 = x.size(0)
torch._assert(x_size_0 == 1)
```

The convert code would try to insert a dequantize after `x_size0`,
because it was a descendant of a quantized node and it was needed
for a non-quantized operation.  Since the actual type of the `size`
function output is an integer, this does not make sense.

This PR fixes the issue by checking if the node has a known integer
return type, and skipping dequantize in this scenario.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_assert_on_size_after_quant_layer
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26780690](https://our.internmc.facebook.com/intern/diff/D26780690)

vkuzo committed:
[not for land yet]: fix using size of quant layer in torch._assert

Summary:

Before this diff, if we had code lik

```
x = any_quant_layer(...)
x_size0 = x.size(0)
torch._assert(x_size_0 == 1)
```

The convert code would try to insert a dequantize after `x_size0`,
because it was a descendant of a quantized node and it was needed
for a non-quantized operation.  Since the actual type of the `size`
function output is an integer, this does not make sense.

For now, this is fixed as a one-off to unblock a customer.  In the
future, we may need to think more deeply about all the functions which
can return non-quantized types from quantized tensors and make sure
they are all covered.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_assert_on_size_after_quant_layer
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land yet]: fix using size of quant layer in torch._assert"


Summary:

Before this diff, if we had code lik

```
x = any_quant_layer(...)
x_size0 = x.size(0)
torch._assert(x_size_0 == 1)
```

The convert code would try to insert a dequantize after `x_size0`,
because it was a descendant of a quantized node and it was needed
for a non-quantized operation.  Since the actual type of the `size`
function output is an integer, this does not make sense.

For now, this is fixed as a one-off to unblock a customer.  In the
future, we may need to think more deeply about all the functions which
can return non-quantized types from quantized tensors and make sure
they are all covered.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_assert_on_size_after_quant_layer
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26780690](https://our.internmc.facebook.com/intern/diff/D26780690)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land yet]: fix using size of quant layer in torch._assert"


Summary:

Before this diff, if we had code lik

```
x = any_quant_layer(...)
x_size0 = x.size(0)
torch._assert(x_size_0 == 1)
```

The convert code would try to insert a dequantize after `x_size0`,
because it was a descendant of a quantized node and it was needed
for a non-quantized operation.  Since the actual type of the `size`
function output is an integer, this does not make sense.

For now, this is fixed as a one-off to unblock a customer.  In the
future, we may need to think more deeply about all the functions which
can return non-quantized types from quantized tensors and make sure
they are all covered.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_assert_on_size_after_quant_layer
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26780690](https://our.internmc.facebook.com/intern/diff/D26780690)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -688,7 +689,12 @@ def load_non_quantized(n: Node) -> Node:
                     str(env) + ' quant_env:' + str(quant_env)
-                env[n.name] = Proxy(quant_env[n.name]).dequantize().node
+                quant_env_node = quant_env[n.name]
+                if node_return_type_is_int(quant_env_node):
```
is it possible to avoid inserting observers for int?

vkuzo committed:
Update on "fx quant: fix using size of quant layer in torch._assert"


Summary:

Before this diff, if we had code lik

```
x = any_quant_layer(...)
x_size0 = x.size(0)
torch._assert(x_size_0 == 1)
```

The convert code would try to insert a dequantize after `x_size0`,
because it was a descendant of a quantized node and it was needed
for a non-quantized operation.  Since the actual type of the `size`
function output is an integer, this does not make sense.

This PR fixes the issue by checking if the node has a known integer
return type, and skipping dequantize in this scenario.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_assert_on_size_after_quant_layer
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26780690](https://our.internmc.facebook.com/intern/diff/D26780690)

[ghstack-poisoned]
